storage/memetcd/peer_store.go

```diff
@@ -0,0 +1,535 @@
+// ------------------------------
+// This method is not atomic for a unique (infohash, peer),
+// so etcd operations can interleave with another invoke of this method for same arguments
+// etcd v3 does not support renew of lease in a transaction which is the primary blocker for above
+// Open github issue: https://github.com/coreos/etcd/issues/8550
+// Because of non-atomicity, method can result in orphaned leases in etcd which are also bound to expire at some point
+// Since tracker will rarely receive concurrent requests for same (infohash, peer), we should do ok with above
+// In-memory locks can also be considered in single active tracker setups but not worth the trouble, in my opinion
+//
+// PERF OPTIMIZATION USING LOCAL TIMESTAMP AS VALUE OF ANNOUNCE KEYS
+// -----------------------------------------------------------------
+// Another odd thing this method does is to put local unix timestamp as value of announce keys
+// which seems counter-intuitive because GC is automatic using leases
+// The reason behind this is to make bootstrapping of a new tracker instance faster
+// by avoiding etcd call to fetch current lease TTL for every announce key
+// If there is significant clock skew across tracker nodes this will result in some keys getting evaluated
+// with incorrect TTL because of local time used as frame of reference
+//
+func (ps *peerStore) etcdPutOrUpdatePeer(ih bittorrent.InfoHash, p bittorrent.Peer, seeder bool) error {
+	current := fmt.Sprintf("%d", timecache.NowUnixNano())
+	key := string(newAnnounceKey(&ih, &p, seeder))
+	ctx, cancel := context.WithTimeout(context.Background(), ps.cfg.EtcdOpTimeout)
+	defer cancel()
+
+	getResp, err := ps.etcd.Get(ctx, key, clientv3.WithLimit(1), clientv3.WithKeysOnly())
+	if err != nil {
+		return fmt.Errorf("failed to do etcd operation, get for %s: %v", key, err)
+	}
+	if getResp.Count == 0 {
+		leaseResp, err := ps.etcd.Grant(ctx, int64(ps.cfg.MemoryStore.PeerLifetime.Seconds()))
+		if err != nil {
+			return fmt.Errorf("failed to do etcd operation, grant lease for %s: %v", key, err)
+		}
+		if _, err = ps.etcd.Put(ctx, key, current, clientv3.WithLease(leaseResp.ID)); err != nil {
+			return fmt.Errorf("failed to do etcd operation, put key-value %s-%s: %v", key, current, err)
+		}
+	} else {
+		if _, err := ps.etcd.KeepAliveOnce(ctx, clientv3.LeaseID(getResp.Kvs[0].Lease)); err != nil {
+			return fmt.Errorf("failed to do etcd operation, renew lease for %s: %v", key, err)
+		}
+		if _, err = ps.etcd.Put(ctx, key, current, clientv3.WithIgnoreLease()); err != nil {
+			return fmt.Errorf("failed to do etcd operation, update key-value %s-%s: %v", key, current, err)
+		}
+	}
+
+	return nil
+}
+
+func (ps *peerStore) etcdDeletePeer(ih bittorrent.InfoHash, p bittorrent.Peer, seeder bool) error {
+	key := string(newAnnounceKey(&ih, &p, seeder))
+	ctx, cancel := context.WithTimeout(context.Background(), ps.cfg.EtcdOpTimeout)
+	defer cancel()
+
+	if _, err := ps.etcd.Delete(ctx, key); err != nil {
```

Comment:
will fix.
I will log this event but not block memory operation. If memory store also does not  have seeder/leecher, it will continue to throw ErrResourceDoesNotExist for caller